### DISCUSSION TITLE
Help getting virtualization working for large lists

### DISCUSSION THREAD

**Discussion by kuraiji** [AvaloniaUI/Avalonia#8173] [Q&A]
[ViewModel](https://github.com/kuraiji/WalkmanPlaylistCreator/blob/master/WalkmanPlaylistCreator/ViewModels/PlaylistCreatorViewModel.cs)
[View](https://github.com/kuraiji/WalkmanPlaylistCreator/blob/master/WalkmanPlaylistCreator/Views/PlaylistCreatorView.axaml)

I have a large list of songs I want to display but using ListBox with WrapPanel doesn't seem to support virtualization. What are my options displaying a list with wrapping style, support for selecting, and virtualization?

**Comment by maxkatz6** [comment]
> ListBox with WrapPanel doesn't seem to support virtualization. What are my options displaying a list with wrapping style, support for selecting, and virtualization

You can use ItemsRepeater with WrapLayout. And items repeater in common scenario has smoother virtualization.
But you will need to implement selection on your own. This control is basically an alternative for ItemsPresenter (which is used inside of ListBox).

I did a bit more advanced SelectionItemsRepeater with multiselection and keyboard navigation support, it basically adds ".selected" style-class to the item, so it can be styled. https://gist.github.com/maxkatz6/285f1cb8798db81b5ad5dfd4daaa768a . I tested and used it only with default StackLayout, but it should work with any type.

Upd: unfortunately WrapLayout is known to have an annoying bug which might be or not be a blocker for you https://github.com/AvaloniaUI/Avalonia/issues/5869 . You might also consider UniformGridLayout as an alternative layout for ItemsRepeater.

**Reply by kuraiji** [reply]
I tried using ItemsRepeater, but it crashes my program. I get a "System.InvalidOperationException: Call from invalid thread" error.

The only thing I changed was the ListBox element calls in the .axaml to just a single 
`<ItemsRepeator Items="{Binding Songs}/>`

**Reply by maxkatz6** [reply]
Well, this error isn't random. Do you update your lists from other threads?
I might be wrong, but items presenter lacks some assertion for a thread, so it would just work there.

**Reply by kuraiji** [reply]
The list is updated on the same ModelView page it is created in by only add operations. The list itself is composed of ModelViews that contain a Model. The model does have a "RxApp.MainThreadScheduler" call to retrieve bitmap data on a seperate thread. Is that the issue?

**Reply by maxkatz6** [reply]
"RxApp.MainThreadScheduler" is the right choice, if reactiveui is set up properly (probably it is, because otherwise it would fail way earlier). 
Easiest way for you is to check the stacktrace and see where invalid access comes from.

**Reply by kuraiji** [reply]
Are you not allowed to add items to a list connected to an ItemRepeater? The stack trace crashes at the step of adding an element to the list. Specifically the Verify Access function in Dispatcher. "Current thread is not the UI thread".

**Reply by maxkatz6** [reply]
You are allowed. But only from UI (main) thread. You need to see where this call goes from stacktrace. If it goes from the task scheduler - that's your problem.

**Reply by kuraiji** [reply]
Moving the .add call to the main thread fixed the crash. I'm not sure how to get ItemsRepeater to display my View though. With ListBox it was automatically able to take the ViewModel and display a View from it. I have no clue how to get that working with ItemsRepeater. This is how it looks right now. 
```
<ItemsRepeater Items="{Binding Songs}">
      <ItemsRepeater.ItemTemplate>
                <DataTemplate DataType="vm:PlaylistCreatorViewModel">
                          <ContentControl>
                              
                          </ContentControl>
                </DataTemplate>
      </ItemsRepeater.ItemTemplate>
      <ItemsRepeater.Layout>
                <UniformGridLayout/>
      </ItemsRepeater.Layout>
</ItemsRepeater>
```

Without ContentControl it would just display the list as text with the datatype name, but adding it makes it completely blank instead. My guess is to add a call to the View between ContentControl like...
`<vm:SongViewModel/>`
The issue is that the ViewModel doesn't have a default constructor, so I'm not sure what to do.

Update: I realized my mistake. I was calling the ViewModel instead of the View. Calling the View fixed the issue.
